protocols/noise/src/keys.rs

```diff
@@ -18,156 +18,185 @@
-            }
-        }
-        Err(NoiseError::InvalidKey)
+impl StaticKeypair {
+    /// Get a reference to the static secret key.
+    pub fn secret(&self) -> &x25519::StaticSecret {
+        &self.secret
     }
 
-    /// Convert this Edwards 25519 public key into a Curve 25519 public key.
-    pub fn into_curve_25519(self) -> PublicKey<Curve25519> {
-        let m = CompressedEdwardsY(self.bytes)
-            .decompress()
-            .expect("Constructing a PublicKey<Ed25519> ensures this is a valid y-coordinate.")
-            .to_montgomery();
-        PublicKey::new(m.0)
+    /// Get a reference to the static public key.
+    pub fn public(&self) -> &StaticPublicKey {
+        &self.public
     }
-}
 
-impl<T> AsRef<[u8]> for PublicKey<T> {
-    fn as_ref(&self) -> &[u8] {
-        &self.bytes
+    /// Create a new static X25519 keypair.
+    pub fn new() -> StaticKeypair {
+        Self::from(x25519::StaticSecret::new(&mut rand::thread_rng()))
     }
-}
 
-/// ECC secret key.
-#[derive(Clone)]
-pub struct SecretKey {
-    scalar: Scalar
+    /// Construct a static X25519 keypair from an Ed25519 keypair.
+    ///
+    /// *Note*: If the Ed25519 keypair is already used in the context
+    /// of other cryptographic protocols outside of Noise, e.g. for
+    /// signing in the `secio` protocol, it should be preferred to
+    /// create a new `StaticKeypair` for use in the Noise protocol.
+    ///
+    /// See also:
+    ///
+    /// [Noise: Static Key Reuse](http://www.noiseprotocol.org/noise.html#security-considerations)
+    /// [Ed25519 to Curve25519](https://libsodium.gitbook.io/doc/advanced/ed25519-curve25519)
+    pub fn from_ed25519(ed: &ed25519::Keypair) -> StaticKeypair {
+        // An Ed25519 public key is derived off the left half of the SHA512 of the
+        // secret scalar, hence a matching conversion of the secret key must do
+        // the same to yield a Curve25519 keypair with the same public key.
+        let ed25519_sk = ed.secret_bytes();
+        let mut curve25519_sk: [u8; 32] = [0; 32];
+        let hash = digest(&SHA512, ed25519_sk);
+        curve25519_sk.copy_from_slice(&hash.as_ref()[..32]);
+        Self::from(x25519::StaticSecret::from(curve25519_sk))
+    }
 }
 
-impl SecretKey {
-    pub(crate) fn new(bytes: [u8; 32]) -> Self {
-        SecretKey {
-            scalar: Scalar::from_bytes_mod_order(bytes)
-        }
+impl From<x25519::StaticSecret> for StaticKeypair {
+    fn from(secret: x25519::StaticSecret) -> StaticKeypair {
+        let public = StaticPublicKey(x25519::PublicKey::from(&secret));
+        StaticKeypair { secret, public }
     }
 }
 
-impl SecretKey {
-    /// Compute the public key from this secret key.
-    ///
-    /// Performs scalar multiplication with Curve25519's base point.
-    pub fn public(&self) -> PublicKey<Curve25519> {
-        PublicKey::new(self.x25519(&X25519_BASEPOINT).0)
-    }
+/// Static X25519 public key.
+pub struct StaticPublicKey(pub x25519::PublicKey);
```

Comment:
Only because the module generally does not intend to hide / encapsulate the use of `x25519-dalek`, but only add a bit of functionality on top. It is nevertheless certainly not really necessary to have the field `pub`. This will probably change anyway in the course of the necessary changes to also support other curves.